File: api/src/validation/menu.ts
import { z } from 'zod';
import { householdIdSchema } from './householdId.js';
import { normalizeThenEnum } from './recipeShared.js';
import { RECIPE_ROLE_VALUES } from '../domain/recipeRoles.js';

/**
 * `Menu.weekStartDate`'s locked SDL type is `AWSDateTime!` even though
 * `menus.week_start_date` is a plain `DATE` column (E2E_MVP_PLAN.md
 * §15.2.4) — a week boundary is a calendar date, deliberately not
 * timezone-attached. This schema accepts the full `AWSDateTime` shape a
 * client sends (`YYYY-MM-DDThh:mm:ss.sssZ`) and derives just the
 * `YYYY-MM-DD` part for the DB parameter — the time-of-day component is
 * never read or stored. Rejects a malformed value with a clear message
 * rather than letting an unparseable date reach `pg` as a raw
 * `22P02`/`22007` error.
 *
 * The calendar-date check re-derives `YYYY-MM-DD` from the parsed `Date`'s
 * own UTC components and compares it back against the input string, rather
 * than only checking `getTime()` for `NaN` — the JS `Date` constructor is
 * overflow-tolerant (`"2026-02-30"` silently rolls forward to March 2
 * instead of producing `NaN`), so a `getTime()`-only check would let an
 * out-of-range calendar date through and on to Postgres as a raw `DATE`
 * literal, defeating the point of validating here at all.
 */
export const weekStartDateSchema = z
  .string()
  .regex(
    /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d{1,3})?(Z|[+-]\d{2}:\d{2})$/,
    'weekStartDate must be a valid AWSDateTime (YYYY-MM-DDThh:mm:ss.sssZ)',
  )
  .transform((value) => value.slice(0, 10))
  .refine(
    (dateOnly) => {
      // The preceding regex already guarantees an all-digit YYYY-MM-DD
      // shape, so this Date is never NaN — only overflow (e.g. "2026-02-30"
      // rolling forward to March 2) is possible here, which this
      // round-trip comparison catches.
      const parsed = new Date(`${dateOnly}T00:00:00.000Z`);
      const year = parsed.getUTCFullYear().toString().padStart(4, '0');
      const month = (parsed.getUTCMonth() + 1).toString().padStart(2, '0');
      const day = parsed.getUTCDate().toString().padStart(2, '0');
      return `${year}-${month}-${day}` === dateOnly;
    },
    { message: 'weekStartDate must be a valid calendar date' },
  );

export const createMenuArgsSchema = z.object({
  householdId: householdIdSchema,
  weekStartDate: weekStartDateSchema,
});

export type CreateMenuArgs = z.infer<typeof createMenuArgsSchema>;

export const menuArgsSchema = z.object({
  householdId: householdIdSchema,
  weekStartDate: weekStartDateSchema,
});

export type MenuArgs = z.infer<typeof menuArgsSchema>;

// Mirrors `shared/schema.graphql`'s `MealType` enum exactly (breakfast,
// lunch, snacks, dinner) — see `validation/updateHouseholdSettings.ts`'s own
// identical `MEAL_TYPES` constant; not shared from a common module today
// since neither file has a natural home for one yet, same duplication this
// codebase already accepted there.
const MEAL_TYPE_VALUES = ['breakfast', 'lunch', 'snacks', 'dinner'] as const;

const mealTypeSchema = normalizeThenEnum(MEAL_TYPE_VALUES);
const recipeRoleSchema = normalizeThenEnum(RECIPE_ROLE_VALUES);

const menuIdSchema = z.string().uuid('menuId must be a valid UUID');
const recipeIdSchema = z.string().uuid('recipeId must be a valid UUID');

/**
 * `dayOfWeek` mirrors `menu_items.day_of_week`'s own `CHECK (day_of_week
 * BETWEEN 0 AND 6)` — validated here too so an out-of-range value surfaces
 * as a typed `ValidationError` before it ever reaches that constraint as a
 * raw `pg` `23514` error.
 */
export const menuItemInputSchema = z.object({
  recipeId: recipeIdSchema,
  dayOfWeek: z.number().int().min(0, 'dayOfWeek must be between 0 and 6').max(6, 'dayOfWeek must be between 0 and 6'),
  mealSlot: mealTypeSchema,
  slotRole: recipeRoleSchema,
  servingsOverride: z.number().int().positive('servingsOverride must be a positive integer').nullish(),
});

export type MenuItemInput = z.infer<typeof menuItemInputSchema>;

export const addMenuItemArgsSchema = z.object({
  menuId: menuIdSchema,
  input: menuItemInputSchema,
});

export type AddMenuItemArgs = z.infer<typeof addMenuItemArgsSchema>;

export const removeMenuItemArgsSchema = z.object({
  id: z.string().uuid('id must be a valid UUID'),
});

export type RemoveMenuItemArgs = z.infer<typeof removeMenuItemArgsSchema>;

/** `Query.autoFillPreview`'s only argument — a pure read, proposes without writing (W10 §16.2.1, D3). */
export const autoFillPreviewArgsSchema = z.object({
  menuId: menuIdSchema,
});

export type AutoFillPreviewArgs = z.infer<typeof autoFillPreviewArgsSchema>;

/**
 * The largest `items` array `autoFillWeek` will accept in one call.
 * `updateHouseholdSettings.ts`'s own `mealStructureEntrySchema` caps each of
 * `carb`/`sabzi_dal`/`accompaniment` at 10 per meal — the theoretical
 * maximum a household could ever configure is 62/day (2 x 3 roles x 10, plus
 * 1 breakfast + 1 snack) x 7 days = 434. 500 is a round, generously-above
 * bound on top of that real ceiling, not an arbitrary guess — closes a
 * defense-in-depth gap security-reviewer flagged: without it, `commitAllItems`
 * would process an unbounded array sequentially, one DB round trip per item,
 * all held under `lockMenu` for the whole call.
 */
const MAX_AUTOFILL_ITEMS = 500;

/**
 * `Mutation.autoFillWeek`'s commit (W10 §16.2.1, D3) — `items` is exactly
 * `MenuItemInput[]`, the same shape a `Query.autoFillPreview` response's
 * `ProposedMenuItem`s echo back, possibly edited by the user via manual
 * swaps before accepting. `overwrite` is required, not nullable — unlike
 * every patch-style optional field elsewhere in this schema, this is a
 * genuine two-valued instruction with no "leave unchanged" reading, so
 * `.nullish()`'s absent-means-unchanged convention does not apply here.
 */
export const autoFillWeekArgsSchema = z.object({
  menuId: menuIdSchema,
  overwrite: z.boolean(),
  items: z.array(menuItemInputSchema).max(MAX_AUTOFILL_ITEMS, `items must contain at most ${MAX_AUTOFILL_ITEMS} entries`),
});

export type AutoFillWeekArgs = z.infer<typeof autoFillWeekArgsSchema>;

/** `Mutation.markMade`'s only argument (W12 S2, E2E_MVP_PLAN.md §18.3 S2). */
export const markMadeArgsSchema = z.object({
  menuItemId: z.string().uuid('menuItemId must be a valid UUID'),
});

export type MarkMadeArgs = z.infer<typeof markMadeArgsSchema>;

File: mobile/lib/features/menu/domain/menu.dart
/// The app's own menu types, independent of GraphQL and of `built_value` —
/// same boundary rule as `features/household/domain/household.dart`. Nothing
/// generated appears here; the mapping from the generated types lives in
/// `../data/menu_mapper.dart`.
library;

import '../../recipes/domain/recipe.dart';
import '../../recipes/domain/recipe_role.dart';

/// One household's meal plan for a single calendar week (E2E_MVP_PLAN.md
/// §15). [weekStartDate] is a calendar date, not a timezone-attached instant
/// — the server stores it as a plain `DATE` and only widens it to
/// `AWSDateTime` on the wire (`docs/E2E_MVP_PLAN.md` §15.2.4); this type
/// keeps the `DateTime` ferry hands back rather than re-narrowing it to a
/// date-only type of its own, since nothing here needs to defend against
/// its time-of-day component (which is always midnight UTC).
class Menu {
  const Menu({
    required this.id,
    required this.householdId,
    required this.weekStartDate,
    required this.items,
  });

  final String id;
  final String householdId;
  final DateTime weekStartDate;
  final List<MenuItem> items;

  /// Every item on [dayOfWeek] (0 = the week's first day), in `Menu.items`'
  /// own server-side order (day, meal slot, placement time) — see
  /// `menuRepository.ts`'s `findMenuItems` for that ordering.
  List<MenuItem> itemsForDay(int dayOfWeek) => items
      .where((MenuItem item) => item.dayOfWeek == dayOfWeek)
      .toList(growable: false);

  @override
  String toString() => 'Menu(id: $id, weekStartDate: $weekStartDate)';

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is Menu && other.id == id;

  @override
  int get hashCode => id.hashCode;
}

/// One recipe placed into one slot of a [Menu]. [slotRole] is captured at
/// placement time and does NOT track [recipe]'s own `role` if that changes
/// later — a menu is a snapshot of what was planned, not a live view
/// (`shared/schema.graphql`'s own `MenuItem` doc).
class MenuItem {
  const MenuItem({
    required this.id,
    required this.menuId,
    required this.recipe,
    required this.dayOfWeek,
    required this.mealSlot,
    required this.slotRole,
    this.servingsOverride,
    this.madeAt,
  });

  final String id;
  final String menuId;
  final Recipe recipe;
  final int dayOfWeek;

  /// Raw `MealType` wire value (`'breakfast'`/`'lunch'`/`'snacks'`/
  /// `'dinner'`) — kept as a string rather than `household/domain/meal_type.dart`'s
  /// `MealType`, which is an encode-only type built for the setup wizard's
  /// own toggles (no `unknown` member, no decode direction — see its own
  /// class doc) and not a fit for a value this type must *decode* from the
  /// server. Same "no consumer needs the typed enum yet" reasoning
  /// `HouseholdSettings.mealsEnabled` already uses.
  final String mealSlot;

  final RecipeRole slotRole;
  final int? servingsOverride;
  final DateTime? madeAt;

  @override
  String toString() =>
      'MenuItem(id: $id, mealSlot: $mealSlot, slotRole: ${slotRole.name})';

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is MenuItem && other.id == id;

  @override
  int get hashCode => id.hashCode;
}

/// The draft `addMenuItem` sends — mirrors `MenuItemInput` (`shared/schema.graphql`)
/// field-for-field, minus nothing: every field on the wire input has a home
/// here.
class NewMenuItem {
  const NewMenuItem({
    required this.recipeId,
    required this.dayOfWeek,
    required this.mealSlot,
    required this.slotRole,
    this.servingsOverride,
  });

  final String recipeId;
  final int dayOfWeek;

  /// Raw `MealType` wire value — see [MenuItem.mealSlot]'s doc for why.
  final String mealSlot;

  final RecipeRole slotRole;
  final int? servingsOverride;
}
